### PR TITLE
fix(rollback): Ignore disabled server groups in automatic rollbacks

### PR DESCRIPTION
**Pull request by luispollo**
We observed that, when a `RollbackClusterStage` was executed to rollback a failed deployment, the `DetermineRollbackCandidatesTask` was including disabled server groups when determining the target server group to rollback to, despite explicit class comments to the contrary. It did filter out disabled server groups when determining the newest server group, but not when actually determining the one to rollback to.

This was easily reproducible by creating 2 server groups `v000` and `v001` in a cluster, then disabling `v001`, then using a pipeline to deploy a new server group `v002` to the cluster. When this deployment failed, instead of rolling back to `v000`, we would rollback to the disabled group `v001`.

The fix should be self-explanatory given the context above.

Update: upon further consideration, we realized that changing this behavior might affect existing pipelines with a `RollbackClusterStage`, and so a condition was added to only exclude disabled server groups if the stage was initiated as part of an automatic rollback from a failed deployment stage.

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy [hunk 1]
```diff
@@ -211,12 +211,12 @@ class MonitoredDeployStrategy implements Strategy {
...
       }
 
-      log.info("Adding `Grow to $p% of Desired Size` stage with context $resizeContext [executionId=${stage.execution.id}]")
+      log.info("Adding `Grow new server group to $p% of Desired Size` stage with context $resizeContext [executionId=${stage.execution.id}]")
 
       def resizeStage = newStage(
         stage.execution,
         ResizeServerGroupStage.TYPE,
-        "Grow to $p% of Desired Size",
+        "Grow new server group to $p% of Desired Size",
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy]: if you are doing this, then you should put the actual ASG name in here, you can get it from
`internalStageData.newServerGroup` (same with the log statement above)

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy [hunk 2]
```diff
@@ -211,12 +211,12 @@ class MonitoredDeployStrategy implements Strategy {
         resizeContext.capacity = savedCapacity // will scale to a percentage of that static capacity
       }
 
-      log.info("Adding `Grow to $p% of Desired Size` stage with context $resizeContext [executionId=${stage.execution.id}]")
+      log.info("Adding `Grow new server group to $p% of Desired Size` stage with context $resizeContext [executionId=${stage.execution.id}]")
```
Comment (luispollo) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy]: This is unrelated. I was just confused with the name of this task/log entry and thought I'd make it clearer.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java [hunk 3]
```diff
@@ -163,28 +163,28 @@ public TaskResult execute(@Nonnull Stage stage) {
             cluster.get("serverGroups"), new TypeReference<List<ServerGroup>>() {});
     serverGroups.sort(Comparator.comparing((ServerGroup o) -> o.createdTime).reversed());
 
+    List<ServerGroup> enabledServerGroups =
```
Comment (gal-yardeni) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: LGTM - so by filtering disabled server groups in `serverGroupsInRegion` , we are making sure to rollback only to an enable server group?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java [hunk 4]
```diff
@@ -163,28 +163,28 @@ public TaskResult execute(@Nonnull Stage stage) {
...
+        serverGroups.stream()
+            .filter(s -> s.disabled == null || !s.disabled)
+            .collect(Collectors.toList());
+
     List<Map> imagesToRestore = new ArrayList<>();
     for (String region : stageData.regions) {
-      List<ServerGroup> serverGroupsInRegion =
-          serverGroups.stream()
+      List<ServerGroup> enabledServerGroupsInRegion =
+          enabledServerGroups.stream()
```
Comment (emjburns) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: I think you could also filter for not null here, and keep the rename to  `enabledServerGroupsInRegion`. It doesn't seem like you use the `enabledServerGroups` list anywhere else.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java (hunk 3 above)
Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: Yep, that's it!

**Comment by luispollo** [comment]
> it seems good to me. but please double-check with red/black

Tested locally with monitored and red/black strategies.

**Comment by luispollo** [comment]
@marchello2000 Before we merge this, I wanted to double check on potential side-effects. I realized the `RollbackClusterStage` is available for users to compose their pipelines, which means there are potentially lots of users who expected/relied on the existing behavior... How do we make sure we're not going to break a ton of pipelines for users inside and outside of Netflix? Is there a way to tell that the stage was added by the user versus the pipeline itself like in the case of deployment failure with auto-rollback?

**Comment by luispollo** [comment]
> @marchello2000 Before we merge this, I wanted to double check on potential side-effects. I realized the `RollbackClusterStage` is available for users to compose their pipelines, which means there are potentially lots of users who expected/relied on the existing behavior... How do we make sure we're not going to break a ton of pipelines for users inside and outside of Netflix? Is there a way to tell that the stage was added by the user versus the pipeline itself like in the case of deployment failure with auto-rollback?

After discussing this with @marchello2000 offline, we agreed the right thing to do would be to keep the old behavior for a rollback stage defined in a user pipeline, and use new behavior only for automatic rollbacks initiated by deployment strategies, to avoid making any assumptions about how users might be using the stage externally.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java (hunk 4 above)
Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: I'm not sure I follow... Filter on not-null on which property, exactly?

Comment (emjburns) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: Ah sorry, filter out the disabled server groups!

Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: This is a different list, though -- the filter here is to take only the enabled server groups in one region.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy (hunk 1 above)
Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy]: Done!

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java (hunk 4 above)
Comment (emjburns) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: never mind!

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy [hunk 5]
```diff
@@ -121,6 +121,11 @@ class CreateServerGroupStage extends AbstractDeployStrategyStage {
       additionalRollbackContext.disableOnly = true
     }
 
+    // When initiating a rollback automatically as part of deployment failure handling, only rollback to a server
+    // group that's enabled, as any disabled ones, even if newer, were likely manually marked so for being "bad"
+    // (e.g. as part of a manual rollback).
+    additionalRollbackContext.disregardDisabledCandidates = true
```
Comment (luispollo) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy]: This is to account for simpler deployment strategies like `RedBlackStrategy`, which don't plan their own rollback stage.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java [hunk 6]
```diff
@@ -165,27 +165,41 @@ public TaskResult execute(@Nonnull Stage stage) {
...
+            moniker.get().getCluster(),
+            region);
         continue;
       }
 
+      ServerGroup newestEnabledServerGroupInRegion = enabledServerGroupsInRegion.get(0);
+      boolean disregardDisabledCandidates =
+          stageData.additionalRollbackContext != null
+              && stageData.additionalRollbackContext.containsKey("disregardDisabledCandidates")
+              && ((Boolean) stageData.additionalRollbackContext.get("disregardDisabledCandidates"))
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: i think you can probably do:
`&& ((Boolean) stageData.additionalRollbackContext.getOrDefault("disregardDisabledCandidates", false))`

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy (hunk 5 above)
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy]: nit: Personally, I prefer something like this:
```suggestion
    additionalRollbackContext.onlyEnabledServerGroups = true
```

(or `excludeDisabledServerGroups`)

Comment (marchello2000) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy]: findImageFromCluster already uses `onlyEnabled` but maybe that's too ambiguous in this context

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java [hunk 7]
```diff
@@ -165,27 +165,41 @@ public TaskResult execute(@Nonnull Stage stage) {
...
+        // no enabled server groups in this region, nothing to rollback from!
+        logger.warn(
+            "No enabled server groups in cluster {} and region {} to rollback from. Skipping this region.",
+            moniker.get().getCluster(),
+            region);
         continue;
       }
 
+      ServerGroup newestEnabledServerGroupInRegion = enabledServerGroupsInRegion.get(0);
+      boolean disregardDisabledCandidates =
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: Another nit here: I find it easier to use positive booleans, e.g. `includeDisabled` or `onlyEnabled`

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java (hunk 6 above)
Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: Yeah, this context switching between Java and Kotlin is getting to me! 😄

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy (hunk 5 above)
Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy]: I went with "disabled" because that's what clouddriver uses in the response:
https://github.com/spinnaker/orca/blob/688bd3d1af8d366f425059c9be89839ffe6a71f3/orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java#L184

Will change to `excludeDisabledServerGroups`

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java (hunk 7 above)
Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java]: Me too, except "include" should be the default in this case when nothing is added to the context...

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy (hunk 5 above)
Comment (luispollo) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy]: Never mind, will take your suggestion.
